Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,129 @@
+1. minimum memory: 500M, maximum memory: 25G
+1. mimimum CPU: 0m, maximum CPU: 24000m
+
+The above requirement can be denoted as a single bucket.
+HVPA controller would divide the "CPU sub-bucket" and the "memory sub-bucket" linearly among the replica counts such that:
+1. each CPU bucket is of same size, ie, `3000m`
+1. each memory bucket is of same size, ie, `(10G - 100M) / 8`or `1.18G`
+
+Following is an image describing the divided CPU bucket. The numbers in the shaded boxes denote the maximum total resources that can be allocated for a replica count.
+![CPU Story 1](./images/CPUStory1.png)
+
+HVPA controller will take the total cpu and memory resources recommended by HPA or VPA, and will scale the application both horizontally and vertically such that:
+1. the total resoureces allocated to application is same as that recommended by HPA or VPA
+1. the total allocated resource is not more than the maximum total resource for that replica count. This way the resource per replica can be capped to a certain maximum value
+
+For example: If an application is running with 2 replicas with 6 cores each, and VPA and HPA recommendations are 12 cores and 4 replicas respectively, then HVPA would scale the application to 3 replicas and 8 cores. This is because both HPA and VPA recommend 24 cores, and this total lies in the 3rd column of 3 replicas. 
+
+#### Story 2
+A user wants to scale his/her application such that:
+1. It scales only vertically with 1 replica till it reaches 1 core and 2GB memory
+1. When more than 2GB memory is required, then the application should be scaled horizontally and vertically till memory reaches 20GB, and CPU reaches 9 cores, until maximum replica count of 8.
+
+In this case, user can provide 2 buckets:
+1. First bucket:
+   * Replicas: minimum: 1, maximum: 1
+   * CPU: minimum: 0, maximum: 1000m
+   * Memory: minimum: 0, maximum: 2G
+1. Second bucket:
+   * Replicas: minimum: 2, maximum: 8
+   * CPU: minimum: 1, maximum: 9000m
+   * Memory: minimum: 2G, maximum: 20G
+
+### Notes/Constraints/Caveats
+
+#### Different recommendations from HPA and VPA
+In such a case, higher recommendations will be used to calculate the desired replica count. This replica count would be used to calculate desired replica size.
+
+#### Different replica count from total CPU and total memory recommendations
+In such a case, higher replica count would be taken as the desired count. Desired memory and CPU per replica would be calculated by dividing desired total memory and total CPU by the chosen replica count respectively.
+
+#### Pod with multiple containers
+In such a case, all the calculations would be done using the total CPU and total memory allocated to all the containers in the pod for which VPA container resource policy is set to "Auto". The total recommended CPU and memory per replica would then be divided in the same proportion as the recommendations by VPA for those containers.
+
+#### User wants to scale only horizontally (TBD)
```

Comment:
Plain HPA? If HVPA, then inside a bucket, you mean? Then min=max until we fall off the cliff on either side?

Reply:
Taking the following example:
If bucket looks like this:
```yaml
minMemory: 5G
maxMemory: 5G
minReplicas: 3
maxReplicas: 6
minCPU: 2
maxCPU: 2
```
HPA wants 2 -> 3 and VPA wants 7Gi -> 9Gi. The aggregated memory recommendations from HPA and VPA are 21G and 18G.
This would put us from previous bucket to this bucket since the total aggregated memory lies in this bucket (between 15G and 30G). 
Case 1: Scaling according to 18Gi:
We would end up with 4 replicas of 9G (9G, because it is max of 9 and 18÷4 = 4.5 )

Case 2: Scaling according to 21Gi:
We would end up with 5 replicas of 9G (9G, because it is max of 9 and 21÷5 = 4.2 )

Case 3:
we handle this case as an exception, and do a reverse calculation. We set the memory value first, and calculate the number of replicas accordingly.
We end up with replicas of size 5G (we also ignore the cache/base requirement of each replica):
Number of replicas according to HPA (21G total): ceil(21÷5) = 5
Number of replicas according to VPA (18G total): ceil(18÷5) = 4
We can go with either 4 or 5 replicas.
In this case, It becomes ~~easier~~ possible to scale only horizontally as long as aggregated memory recommendations are inside this bucket.

However, if we take into account aggregated CPU recommendation also, it becomes further complicated. It may not necessarily put us in the same bucket.
We can choose to stick with case 1 or 2 above for CPU size, but then we would not be scaling only horizontally if CPU recommendations change.
If we choose to go ahead with case 3, and fix the CPU to 2 cores, we might end up with blocking lot more CPU than required.

Goal here is to be able to scale an application only horizontally under certain situations. It should be possible by some bucket configuration - either by using single bucket, or multiple buckets.
As discussed during the call also, since we don't need this "feature" for our specific case of kube-apiserver scaling, and is more towards a general implementation of HVPA controller, we can decide either way for now.

Reply:
To summarise:
In such a situation, HVPA controller:
1. Sets memory to 5GB, even if VPA recommends more
    Effects: The points mentioned under `Base usage of each replica` are ignored
1. Sets CPU equal to 2 cores, even if VPA recommendations are less
    Effects: May ends up blocking lot more CPU than requested.

Also, VPA tuning to have no affect as long as any of the aggregate recommendation - cpu, memory - fall in such a bucket

Reply:
Yes. But there must be some mechanism in the spec for the user to specify that the scaling in that bucket must be horizontal only. Could you propose something for this?

Reply:
How about this:
```
    - cpu:
        minValue: "2",
        maxValue: "2",
      memory:
        minValue: "5G",
        maxValue: "5G",
      replicas:
        minValue: "6",
        maxValue: "3"
```
Basically if min and max for cpu and memory are same, then HVPA would scale only horizontally, and user will have to live with the implications as mentioned in the comment.

Reply:
Seems ok for the horizontal only case. But I don't like specifying min and max for other scenarios where vertical scaling is involved. How would be the spec then? Can you please give the full picture for the spec?

Reply:
```yaml
spec:
  scaleIntervals:
    - cpu:
        minValue: "100m",
        maxValue: "3",
      memory:
        minValue: "100M",
        maxValue: "7.5G",
      replicas:
        minValue: "1",
        maxValue: "2",
    - cpu:
        minValue: "2",
        maxValue: "2",
      memory:
        minValue: "5G",
        maxValue: "5G",
      replicas:
        minValue: "3",
        maxValue: "6",
```
We don't really need `minValue` to be defined in any of the case because it can be calculated using previous bucket's maxValue. I have updated the doc, and marked it "optional" in the latest commit.
I would still like to keep it in the spec, and use it in implementation for easy readability even if not provided by the user. This field could even help in keeping the whole logic flexible in case we need to change it in future. Example: No spec change required if we don't want "overlapping" buckets [1]. I am not strongly opinionated on this though, and open to suggestions.

[1]: Overlapping buckets: When a bucket's maxValue is more than next bucket's minValue

Reply:
Making it optional is not ideal. That means user might still specify it and that means that we should validate that there are not overlaps or gaps. And if we allow overlaps then we have to specify what the behaviour is if there are overlaps or gaps.

Reply:
makes sense. will remove `minValue`.

Reply:
In the last call, we discussed about having some kind of hysteresis around bucket boundaries.
Should we reintroduce `minValue` to achieve that, or can we have that in any other way?
Open to suggestions. :)

@amshuman-kr @vlerenc

Reply:
See the out-of-band document. It had min values, but we can also go with a formula if well defined, but I fear that's not a good approach as we don't know what a good formula should look like and whether it even exists. Changing the formula, is a change in behaviour, so explicit fields are preferred while we are experimenting and looking for the optimum.

In other words: clear recommendation for explicit (min) fields.

What I also miss is the exact description of other points, e.g.:
> Scale-Up & Out:
> Current Requests: 150m / 1.8Gi with 1 replica
> VPA Recommendation: 150m / 2.2Gi
> New Requests: 150m / 2.2Gi with 1->2 replicas, because 2.2Gi lands us in the “second” bucket when out-scaling that has 2 replicas; we carry over the vertical scaling as is, which is most conservative.
> Notice: New CPU would keep us in bucket 1 while memory brings us to bucket 2, so we go to bucket 2.
> Reasoning: We don’t know what the cache percentage is and if it is very high, the new replica will need almost the same memory to hold the cache. If not, VPA will eventually scale down.
 
> Scale-In & Down:
> Current Requests: 150m / 2.2Gi with 3 replicas
> VPA Recommendation: 150m / 1.8Gi
> New Requests: 225m / 2.7Gi with 3->2 replicas, because 1.8Gi lands us in the “second” bucket when in-scaling that has 2 replicas; we adapt the vertical scaling based on the replica ratio, which is most conservative (3 * 1.8Gi = 5.4 Gi overall memory requests from the PoV of VPA; 5.4 Gi / 2 = 2.7 Gi memory requests per replica that preserves this value; CPU the same, i.e. 3 / 2 * 150m = 225m)
> Notice: Old and new CPU would even land us in bucket 1, but memory brings us only down to bucket 2, so we go to bucket 2 with 2 replicas.
> Reasoning: We don’t know what the cache percentage is and if it is very low, the remaining replicas will need even more memory to cope with the requests. If not, VPA will eventually scale down.
 
> Question: In the very special case of the API server (which we are most concerned with), we could also decide to scale horizontally only based on CPU usage and ignore memory altogether. If the existing API server replicas require more memory, VPA will get them more, but if the CPU load increases beyond the bucket upper limit for CPU, then and only then we change the bucket. WDYT? Same for scale-in?

Where is the description of that behaviour described in detail, i.e. how HVPA should behave in transition, up/out <-> down/in.

Reply:
Added the exact examples under section `kube-apiserver scaling example` in https://github.com/gardener/hvpa-controller/commit/5e0df99420af1122a36a6b66d3484078562bdbb0.
Hope this suffices.

>In other words: clear recommendation for explicit (min) fields.

I think @amshuman-kr has some other ideas on how hysteresis can be achieved without min fields as per previous comments in this thread, specifically:
>But I don't like specifying min and max for other scenarios where vertical scaling is involved

IMO too, easiest would be to have the min fields.
Let's wait to hear from him to understand the approach that he might have in mind before making any changes to prevent going back and forth (again). :)

Reply:
I have added options to achieve hysteresis/stability at bucket boundaries in https://github.com/gardener/hvpa-controller/commit/883a938ffd07ddfd2525c171afa7e676974775d8.
I have also added examples for each of the options to explain how they will work.
PTAL. Thanks.